### ISSUE TITLE
Create the new Collation RLP object

### ISSUE THREAD

**Issue by hwwhww** [ethereum/py-evm#420]
### What is wrong?
(Updated the issue, separated this ticket from VMC contract)

#### `CollationHeader`
```
shard_id           uint256 // pointer to shard
parent_hash        bytes32 // pointer to parent header
chunks_root        bytes32 // pointer to collation body
period             uint256 // timestamp
height             uint256 // height in collation tree
proposer_address   address
proposer_bid       uint256
proposer_signature bytes
```

#### `Collation`
```
header     CollationHeader
blobs        list of blob
```

### How can it be fixed

Since there are two other issues are related to VMC, it might be better to create a new `Collation` class to avoiding influencing VMC.

**Comment by mhchia** [comment]
I think I can work on this.
Change the collation structure according to the new spec, and do VMC optimization

**Comment by hwwhww** [comment]
@mhchia 
1. Some fields renamings are still in discussion.
2. Removing `state_root` of `Collation` may make some tests fails due to this line: https://github.com/ethereum/py-evm/blob/3784a87d0850c92d5865d97ce4bb6f5075d47896/evm/vm/forks/sharding/vm_state.py#L45
    Which means we *may* have to make another *temporary* `NewCollation` class **or** decouple state execution.

I suggest something you can do before this issue:
1. Move VMC out of `evm/vm` because (1) it's not part of VM and (2) help for decoupling state execution and other processes.
2. Ping @pipermerriam , what do you think of we move VMC contract to `ethereum/sharding` repository and we make sharding repository to be a submodule of py-evm? If it's okay, @mhchia would you help to move?
2. Write down basic VMC API doc. (don't need to be perfect)

**Comment by mhchia** [comment]
@hwwhww Thank you for the suggestion. I would like to help. But I think I should finish `GUESS_HEAD` first, and then come to this issue after. So sorry for this.
If anyone would like to do this please feel free to take it.

**Comment by jannikluhn** [comment]
This seems to overlap with #436

**Comment by hwwhww** [comment]
@jannikluhn A little bit, but my plan is that  #420 defines the `CollationHeader` and the blobs in `CollationBody` are just binary data; #436 handles the blobs encoding and the binary chunk Merkle tree.

**Comment by hwwhww** [comment]
Updated to ticket description.

**Comment by mhchia** [comment]
@jannikluhn 
 > Some fields have type uint256, which I assume means that encoded integers are padded with zeros to a size of 32 bytes. This makes sense for the SMC (easier to decode) but not so much for RLP encoding (takes unnecessary space). Can we encode integers for the RLP format in the traditional way (like block headers are today) and then provide some kind of "encode for SMC" method?

If we do this, it seems
1. we will need to implement the minimal rlp decoder for `big_endian_int`. I'm not sure if it is worthwhile or not.
2. Or we just use the rlp decoder contract. But it needs an additional contract call.

> Related: What's the hash of a collation header? A block header hash is keccak(rlp(header)), in the existing collation implementation it's keccak(shard_id ++ period ++ ...). Do we want to have both, should the SMC learn RLP, or should we go with the field-concatenated hash everywhere?

IMO for simplicity, using field-concatenated hash seems easier. I'm not sure if it is worthwhile to make SMC able to RLP encode. We will have to implement it and test it further. And maybe it will cost more gas to do rlp encoding compared to the field-concatenated hash.

> There's no reference to the main chain anymore (in the previous version there used to be a "period_start_prevhash"). Is that intentional?

I'm not sure of it either. Maybe need to ask @hwwhww 

> Is there any format for the proposer signature? I think at least a maximum size would be necessary, but we could also go further and have v/r/s integers just like (old-style) transactions (although I kind of like having a single signature field).

I think we can use the previous format of signature. It is the just the concatenation of `v`, `r`, `s`. (https://github.com/ethereum/sharding/blob/develop/old_sharding_poc/sharding/contract_utils.py#L19-L22). And it can be verified through the [validation code](https://github.com/ethereum/sharding/blob/develop/old_sharding_poc/sharding/validator_manager_utils.py#L44-L59) we used before.

> What's the proposer signature exactly? Same as transaction signature, but without the replay protection hacks?

I don't know it for now either. It is not currently included in the spec and need to be clarified as well. If we want to do it now, maybe we can start off letting the `sig_hash_of_proposer` to be `keccak(concat(shard_id ++ parent_hash ++ ... ++ proposer_bid))` or something like this.

Just some thoughts. Please correct me if there is anything wrong :)
And we can discuss about these on the meeting tomorrow.

**Comment by hwwhww** [comment]
> There's no reference to the main chain anymore (in the previous version there used to be a"period_start_prevhash"). Is that intentional?

yes!

Also, we're thinking about removing `height` too since I couldn't find a scenario of why we definitely *need* height in header ... I can say it's convenient in some cases, but now we're thinking of just log `height` in the EVM log or just query the `score` from SMC (in JMRS, `score` still equals `height`).

```
shard_id           uint256 // pointer to shard
parent_hash        bytes32 // pointer to parent header
chunks_root        bytes32 // pointer to collation body
period             uint256
proposer_address   address
proposer_bid       uint256
proposer_signature bytes
```

**Comment by hwwhww** [comment]
closed via #462
